docs/proposals/29-autoscaling-storage-volumes.md

```diff
@@ -0,0 +1,326 @@
+
+### Limitations
+#### Reliance on Prometheus
+`pvc-autoscaler` relies on metrics from the seed's cache Prometheus instance. The availability of the otherwise simple,
+and lightly and predictably loaded, autoscaler is thus limited to the availability level of the complex Prometheus, which
+may be subject to load spikes and autoscaling evictions. By itself, this limitation is deemed acceptable, as the trigger
+thresholds, configured for PVC autoscaling, usually afford tolerance to substantial delays.
+
+Using the seed's cache Prometheus instance creates a dependency loop - `pvc-autoscaler` needs `cache-prometheus` in order
+to perform scaling, and `cache-prometheus` needs `pvc-autoscaler` to scale its database volume. That dependency results
+in one mode of persistent failure:
+1. `cache-prometheus` exhausts its storage before `pvc-autoscaler` resizes it
+2. `cache-prometheus` enters a fault loop, due to inability ot expand database
+3. `pvc-autoscaler` enters fault loop, due to driving metrics signal unavailable
+4. At this point, autoscaling seizes for all PVs on the seed
+
+To mitigate the risk of such failure, a larger initial storage request is used for `cache-prometheus` - 6Gi. It offers a
+1Gi headroom over the 5Gi storage retention limit, configured on `cache-prometheus`. Field data suggests 1Gi should
+suffice to accommodate the occasional usage spikes which exceed the data retention limit.
+
+#### Latency
+The pipeline which propagates the volume metrics, driving `pvc-autoscaler`, is as follows:
+1. `kubelet` collects PV state data and publishes resource metrics. Default poll interval: 1 minute
+   ([source](https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/))
+2. `metrics-server` polls `kubelet` resource metrics. Default poll interval: 1 minute
+   ([source](https://github.com/kubernetes-sigs/metrics-server/blob/master/FAQ.md))
+3. `kube-apiserver` forwards metrics API calls to `metrics-server`, which is the registered APIService for metrics. Call
+   forwarding is instantaneous, so the delay is minimal.
+4. The seed's cache Prometheus polls the metrics provided by `kube-apiserver`. Default poll interval: 1 minute.
+5. `pvc-autoscaler` polls the seed's cache Prometheus instance. Poll interval: 1 minute.
+
+The reaction time, until `pvc-autoscaler` **initiates** a volume resize, can thus exceed 4 minutes. The resize operation
+itself, including the file system resize, can further take a few minutes. The cumulative delay, roughly estimated to up
+to 10 minutes, is mitigated via an appropriate safety margin, provided by the choice of utilisation level, configured
+as a scaling trigger threshold.
+
+## References
+- [[1]] `pvc-autoscaler` GitHub repo
+- [[2]] `pvc-autoscaler` evaluation of existing autoscalers
+- [[3]] `kube-rbac-proxy` GitHub project
+- [[4]] Kubernetes TokenReview API
+- [[5]] Kubernetes SubjectAccessReview API
+
+[1]: https://github.com/gardener/pvc-autoscaler
```

Comment:
`pvc-autoscaler` repo is still private. Is it required for the purpose of this review?

Reply:
Formally - no, but with the repo private, the GEP would be suggesting the use of some obscure, completely undocumented autoscaler.

So if we feel that the `pvc-autoscaler` repo is already good enough to publish, that may be the easiest path. Next option is, publish just the binary and documentation. Last option (still viable): publish just the binary and live with consuming an obscure application for a while.